src/clustering_quality.jl

```diff
@@ -0,0 +1,316 @@
+    n == k && throw(ArgumentError("Quality index not defined for the degenerated clustering where each data point is its own cluster."))
+    all(>=(0), weights) || throw(ArgumentError("All weights must be larger or equal 0."))
+    1 < fuzziness || throw(ArgumentError("Fuzziness must be greater than 1 ($fuzziness given)"))
+
+    if quality_index == :calinski_harabasz
+        _cluquality_calinski_harabasz(metric, data, centers, weights, fuzziness)
+    elseif quality_index == :xie_beni
+        _cluquality_xie_beni(metric, data, centers, weights, fuzziness)
+    else
+        throw(ArgumentError("Quality index $quality_index not supported."))
+    end
+end
+
+clustering_quality(data::AbstractMatrix{<:Real}, R::FuzzyCMeansResult; quality_index::Symbol, fuzziness::Real, metric::SemiMetric=SqEuclidean()) =
+    clustering_quality(data, R.centers, R.weights; quality_index = quality_index, fuzziness = fuzziness, metric = metric)
+
+
+# interface of clustering indices with cluster centres not known
+
+function clustering_quality( 
+        assignments::AbstractVector{<:Integer}, # n vector
+        dist::AbstractMatrix{<:Real};           # n×n matrix
+        quality_index::Symbol 
+    )
+    n, m = size(dist)
+    na = length(assignments)
+    n == m || throw(ArgumentError("Distance matrix must be square."))
+    n == na || throw(DimensionMismatch("Inconsistent array dimensions for distance matrix and assignments."))
+
+    if quality_index == :silhouettes
+        mean(silhouettes(assignments, dist))
+    elseif quality_index == :dunn
+        _cluquality_dunn(assignments, dist)
+    else
+        throw(ArgumentError("Quality index $quality_index not supported."))
+    end
+end
+
+
+clustering_quality(data::AbstractMatrix{<:Real}, assignments::AbstractVector{<:Integer}; quality_index::Symbol, metric::SemiMetric=SqEuclidean()) = 
+    clustering_quality(assignments, pairwise(metric, data, dims=2); quality_index = quality_index)
+
+clustering_quality(data::AbstractMatrix{<:Real}, R::ClusteringResult;  quality_index::Symbol, metric::SemiMetric=SqEuclidean()) =
+    clustering_quality(R.assignments, pairwise(metric, data, dims=2); quality_index = quality_index)
+
+clustering_quality(R::ClusteringResult, dist::AbstractMatrix{<:Real}; quality_index::Symbol) = 
+    clustering_quality(R.assignments, dist; quality_index = quality_index)
+
+
+# utility functions
+
+function _gather_samples(assignments, k) # cluster_samples[j]: indices of points in cluster j
+    cluster_samples = [Int[] for _ in  1:k]
+    for (i, a) in zip(eachindex(assignments), assignments)
+        push!(cluster_samples[a], i)
+    end
+    return cluster_samples
+end
+
+# shared between hard clustering calinski_harabasz and xie_beni
+function _inner_inertia(
+        metric::SemiMetric,
+        data::AbstractMatrix,
+        centers::AbstractMatrix,
+        assignments::AbstractVector{<:Integer}
+    ) 
```

Comment:
To support clustering type-invariant version of `_cluquality_xie_beny()`
```suggestion
        assignments::AbstractVector{<:Integer},
        fuzziness::Nothing
    ) 
```